### PR TITLE
:zap: Support Scalars and minor fixes

### LINKED ISSUE #149: Support multiplying by numbers
Create the unit tensor right before the forward pass

https://github.com/geohot/tinygrad/blob/0cf21881b710dd55614a1eb19c5ca0e7081cb8b5/tinygrad/tensor.py#L215

### PR DESCRIPTION
**Pull request by kartik4949**
fixes #149

### CODE REVIEW

**Comment by kartik4949** [comment]
@geohot can i add right operations in this pr?
i.e __radd__, __rmul__
because its important :)

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -245,13 +243,17 @@ def register(name, fxn, gpu=False):
...
+    x = tuple((Tensor(arg,requires_grad=False) if not isinstance(arg, Tensor) else arg for arg in x ))
     f = (Tensor.opsgpu if x[0].gpu else Tensor.ops)[name]
     f.cl_ctx, f.cl_queue = cl_ctx, cl_queue
     return f.apply(f, *x, **kwargs)
   setattr(Tensor, name, dispatch)
   if name in ['add', 'sub', 'mul', 'div', 'pow']:
     setattr(Tensor, "__%s__" % name, dispatch)
     setattr(Tensor, "__i%s__" % name, lambda self,x: self.assign(dispatch(self,x)))
+    if "__r%s__" % name in dir(int):
+        setattr(Tensor, "__r%s__" % name, lambda self, x: getattr(self, "__%s__" % name)(x))
```
Comment (kartik4949) [inline tinygrad/tensor.py:255]: this because of line constraints :(

File: tinygrad/ops.py [hunk 2]
```diff
@@ -7,6 +7,8 @@
 def unbroadcast(out, in_sh):
   # adjoint operation to broadcast is sum. Need to sum all axis with 1 = in_sh[i] < out.shape[i]
   sum_axis = [i for i in range(len(in_sh)) if in_sh[i]==1 and out.shape[i]>1]
+  if not in_sh: # incase input is a scalar.
```
Comment (kartik4949) [inline tinygrad/ops.py:10]: a hackish way... i know ..
im not sure if its readable 
@geohot can u suggest something here?

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -245,13 +243,17 @@ def register(name, fxn, gpu=False):
   else:
     Tensor.ops[name] = fxn
   def dispatch(*x, **kwargs):
+    x = tuple((Tensor(arg,requires_grad=False) if not isinstance(arg, Tensor) else arg for arg in x ))
```
Comment (kartik4949) [inline tinygrad/tensor.py:246]: will add gpu=x[0].gpu

**Comment by geohot** [comment]
Looking good!

Absolutely yes on right operations.

Though we shouldn't be playing code golf. If you want more lines, refactor something else. If it's not the cleanest way to write it, don't. I will merge even if it breaks the line limit if it's clean, then I'll save the lines myself.

File: tinygrad/ops.py (hunk 2 above)
Comment (geohot) [reply tinygrad/ops.py:10]: Oh, this shouldn't be in ops

File: tinygrad/tensor.py [hunk 4]
```diff
@@ -245,13 +243,17 @@ def register(name, fxn, gpu=False):
...
     f = (Tensor.opsgpu if x[0].gpu else Tensor.ops)[name]
     f.cl_ctx, f.cl_queue = cl_ctx, cl_queue
     return f.apply(f, *x, **kwargs)
   setattr(Tensor, name, dispatch)
   if name in ['add', 'sub', 'mul', 'div', 'pow']:
-    setattr(Tensor, "__%s__" % name, dispatch)
-    setattr(Tensor, "__i%s__" % name, lambda self,x: self.assign(dispatch(self,x)))
+    setattr(Tensor, f"__{name}__", dispatch)
+    setattr(Tensor, f"__i{name}__", lambda self,x: self.assign(dispatch(self,x)))
+    if "__r%s__" % name in dir(int):
```
Comment (geohot) [inline tinygrad/tensor.py:254]: Which does this fail for?

File: tinygrad/tensor.py [hunk 5]
```diff
@@ -59,7 +59,7 @@ class Tensor:
   def __init__(self, data, gpu=None, requires_grad=True):
     if gpu is None:
       gpu = Tensor.default_gpu
-    if isinstance(data, list):
+    if isinstance(data, (list,float)):
```
Comment (geohot) [inline tinygrad/tensor.py:62]: np.array[data]) for float?

File: tinygrad/tensor.py [hunk 6]
```diff
@@ -208,9 +208,7 @@ def swish(self):
     return self.mul(self.sigmoid())
 
   def tanh(self):
-    t2 = Tensor(np.zeros(self.shape, dtype=self.dtype)+2, gpu=self.gpu, requires_grad=False)
-    t1 = Tensor(np.zeros(self.shape, dtype=self.dtype)+1, gpu=self.gpu, requires_grad=False)
-    return self.mul(t2).sigmoid().mul(t2) - t1 # 2*sigmoid(2*x)-1
+    return 2.0 * ((2.0 * self).sigmoid())  - 1.0 # 2*sigmoid(2*x)-1
```
Comment (geohot) [inline tinygrad/tensor.py:211]: I see 3 lines to save in the other ops!

File: tinygrad/tensor.py (hunk 4 above)
Comment (kartik4949) [reply tinygrad/tensor.py:254]: "__rdiv__"

File: tinygrad/tensor.py (hunk 5 above)
Comment (kartik4949) [reply tinygrad/tensor.py:62]: sure

File: tinygrad/tensor.py (hunk 6 above)
Comment (kartik4949) [reply tinygrad/tensor.py:211]: @geohot i saved them in mean, sqrt but got error in gpuops :(
can u help me debug it.

File: tinygrad/ops.py (hunk 2 above)
Comment (kartik4949) [reply tinygrad/ops.py:10]: well @geohot,
if input is scalar i.e 2.0 * Tensor([1.0,2.0])
this 2.0 scalar is converted to a numpy array with shape (),
but during multiplication in 

> Looking good!
> 
> Absolutely yes on right operations.
> 
> Though we shouldn't be playing code golf. If you want more lines, refactor something else. If it's not the cleanest way to write it, don't. I will merge even if it breaks the line limit if it's clean, then I'll save the lines myself.

point noted :) @geohot

File: tinygrad/tensor.py (hunk 1 above)
Comment (kartik4949) [reply tinygrad/tensor.py]: @geohot this looks good, its already readable.

File: tinygrad/ops.py (hunk 2 above)
Comment (kartik4949) [reply tinygrad/ops.py:10]: @geohot for scalar multiplications i.e 
Tensor object with multidimensional data * 2.0
this 2.0 is converted to numpy.array with shape = ()
but due to broadcasting this numpy changes to dimension of Tensor for multiplication

and for backward pass  inside broadcasting y.shape is still = (), and out_grad is mulltidimensional hence  out_grad wont be compatible with y as it just should be scalar. so i did , out.flat[0] which basically picks first element. :)

**Comment by marcelbischoff** [comment]
I had a similar issue with pow, had to reshape so that backward works

File: tinygrad/ops.py (hunk 2 above)
Comment (geohot) [reply tinygrad/ops.py:10]: Convert it to a numpy array with shape (1,) instead.

File: tinygrad/tensor.py (hunk 6 above)
Comment (geohot) [reply tinygrad/tensor.py:211]: Hmm, I'm not sure tanh is tested...paste the error

File: tinygrad/tensor.py (hunk 4 above)
Comment (geohot) [reply tinygrad/tensor.py:254]: Hmm, I think that's a different problem.

https://stackoverflow.com/questions/34344025/have-the-rdiv-and-idiv-operators-changed-in-python-3-x

File: tinygrad/tensor.py (hunk 6 above)
Comment (kartik4949) [reply tinygrad/tensor.py:211]: tanh is tested

File: tinygrad/ops.py (hunk 2 above)
Comment (kartik4949) [reply tinygrad/ops.py:10]: cannot reshape (2925,) to (1,)

File: tinygrad/tensor.py (hunk 4 above)
Comment (kartik4949) [reply tinygrad/tensor.py:254]: yeah we dont support right div in here

File: tinygrad/tensor.py (hunk 6 above)
Comment (kartik4949) [reply tinygrad/tensor.py:211]: E     TypeError: __init__(): incompatible constructor arguments. The following argument types are supported:
E         1. pyopencl._cl.Buffer(context: pyopencl._cl.Context, flags: int, size: int = 0, hostbuf: object = None)
E     
E     Invoked with: <pyopencl.Context at 0x557e1da84a40 on <pyopencl.Device 'GeForce GTX 1650' on 'NVIDIA CUDA' at 0x557e1e3d09a0>>, 33, 4.0; kwargs: hostbuf=array([1.], dtype=float32)

tinygrad/tensor.py:45: TypeError

Comment (kartik4949) [reply tinygrad/tensor.py:211]: when changed
```
def mean(self)
   return self.sum() * (1/np.prod(self.shape))


```

Comment (kartik4949) [reply tinygrad/tensor.py:211]: @geohot  this is solved :)

**Comment by kartik4949** [comment]
> I had a similar issue with pow, had to reshape so that backward works

@marcelbischoff can u share what u changed :)

**Comment by kartik4949** [comment]
@geohot 
PR is completed
1. tanh gpu testing added
2. unbroadcasting for scalars bug fixed for both (cpu and gpu)
3. right operations, rtruedivide , itruedivide and truedivide
4. reduced first class ops lines :)

for unbroadcasting cpu code, if u have a better way for scalar grad please change:)

File: tinygrad/ops.py (hunk 2 above)
Comment (kartik4949) [reply tinygrad/ops.py]: @geohot got it , just need to change sum_axis = None :)

**Comment by geohot** [comment]
There's merge conflicts

File: tinygrad/ops.py [hunk 7]
```diff
@@ -7,7 +7,11 @@
 def unbroadcast(out, in_sh):
   # adjoint operation to broadcast is sum. Need to sum all axis with 1 = in_sh[i] < out.shape[i]
   sum_axis = [i for i in range(len(in_sh)) if in_sh[i]==1 and out.shape[i]>1]
-  return out.sum(axis=tuple(sum_axis)).reshape(in_sh)
+  sum_axis = tuple(sum_axis)
```
Comment (geohot) [inline tinygrad/ops.py:10]: This shouldn't be here. Scalars shouldn't modify ops.

File: tinygrad/tensor.py [hunk 8]
```diff
@@ -245,13 +240,22 @@ def register(name, fxn, gpu=False):
...
   def dispatch(*x, **kwargs):
+    x = tuple((Tensor(np.array([arg]), gpu=x[0].gpu, requires_grad=False) if not isinstance(arg, Tensor) else arg for arg in x ))
     f = (Tensor.opsgpu if x[0].gpu else Tensor.ops)[name]
     f.cl_ctx, f.cl_queue = cl_ctx, cl_queue
     return f.apply(f, *x, **kwargs)
   setattr(Tensor, name, dispatch)
   if name in ['add', 'sub', 'mul', 'div', 'pow']:
-    setattr(Tensor, "__%s__" % name, dispatch)
-    setattr(Tensor, "__i%s__" % name, lambda self,x: self.assign(dispatch(self,x)))
+    if "__%s__" % name in dir(int):
```
Comment (geohot) [inline tinygrad/tensor.py:249]: truepow? This should be explicit, not if in dir(int)

File: tinygrad/tensor.py [hunk 9]
```diff
@@ -193,24 +193,19 @@ def detach(self):
...
+    return self.pow(0.5)
 
   def div(self, y):
-    root = Tensor(np.zeros(self.shape, dtype=self.dtype)-1, gpu=self.gpu, requires_grad=False)
-    return self.mul(y.pow(root))
+    return self * (y**-1.0)
 
   def swish(self):
-    return self.mul(self.sigmoid())
+    return self * (self.sigmoid())
```
Comment (geohot) [inline tinygrad/tensor.py:205]: Extra parens

File: tinygrad/ops.py (hunk 7 above)
Comment (geohot) [reply tinygrad/ops.py:10]: Hmm, actually this is a bug in the ops, you are right.

**Comment by geohot** [comment]
Gonna clean up and merge. It's right, just code clean ups.

**Comment by geohot** [comment]
https://github.com/geohot/tinygrad/commit/c4540f1b8cd6f98a2f97d601bf5bd2a87c0b86d4

The issue with div was that register is never called with it, since it's a second class op.

TODO also, update the optimizers to use scalars instead of "one" and "two"

**Comment by kartik4949** [comment]
@geohot think about this
for scalars we dont have to calculate gradient right.
we can bypass gradientcalculation in backward pass if its scalar, it can save some time in unnecessary gradient calculation where requires_gradient is false. 
what u think?

e.g
```
x = Tensor([2.0,2.0,2.0]) * 3.0

class Mul():
  def backward(ctx, grad_out):
      x, y = ctx.saved_backward
      if y is scalar:
          return y*grad_out, gradient_for_y_will_be_arbitary_number
     return y*grad_out, grady 
```   
this can help especially in **GPU backward**  as reduce_op is very costly :)
thoughts?

**Comment by marcelbischoff** [comment]
that should also fix #155 (at least for me)

**Pull request closed without merging**